File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/payment/application/query/GetPaymentService.java
package com.example.accountservice.payment.application.query;

import com.example.accountservice.payment.domain.Payment;
import com.example.accountservice.payment.domain.PaymentException;
import com.example.accountservice.payment.domain.PaymentFindQuery;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class GetPaymentService {

    private final PaymentQuery paymentQuery;

    public GetPaymentResult getPayment(String paymentId, String requesterId) {
        Payment payment =
                paymentQuery
                        .findPayments(new PaymentFindQuery(0, 1, paymentId, requesterId))
                        .payments()
                        .stream()
                        .findFirst()
                        .orElseThrow(
                                () ->
                                        new PaymentException(
                                                PaymentException.ErrorCode.PAYMENT_NOT_FOUND,
                                                "Payment not found."));
        return GetPaymentResult.from(payment);
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/payment/application/query/GetPaymentService.kt
package com.example.accountservice.payment.application.query

import com.example.accountservice.payment.domain.PaymentFindQuery
import com.example.accountservice.payment.domain.PaymentNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class GetPaymentService(
    private val paymentQuery: PaymentQuery,
) {
    fun getPayment(
        paymentId: String,
        requesterId: String,
    ): GetPaymentResult {
        val (payments, _) =
            paymentQuery.findPayments(
                PaymentFindQuery(page = 0, take = 1, paymentId = paymentId, ownerId = requesterId),
            )
        val payment = payments.firstOrNull() ?: throw PaymentNotFoundException(paymentId)
        return GetPaymentResult(
            paymentId = payment.paymentId,
            cardId = payment.cardId,
            accountId = payment.accountId,
            ownerId = payment.ownerId,
            amount = payment.amount,
            status = payment.status.name,
            createdAt = payment.createdAt,
        )
    }
}
